custom_components/localtuya/pytuya/__init__.py

```diff
@@ -170,51 +202,56 @@ def __init__(
         self.version = protocol_version
         self.dev_type = "type_0a"
         self.dps_to_request = {}
+        self.cipher = AESCipher(self.local_key)
+        self.seqno = 0
 
         self.port = 6668  # default - do not expect caller to pass in
 
-    def __repr__(self):
-        """Return internal string representation of object."""
-        return "%r" % ((self.id, self.address),)  # FIXME can do better than this
-
-    def _send_receive(self, payload):
-        """
-        Send single buffer `payload` and receive a single buffer.
+    def exchange(self, command, data=None):
+        """Send and recive a message, returning response from device."""
+        _LOGGER.debug("Sending command %s (device type: %s", self.dev_type)
+        payload = self._generate_payload(command, data)
+        dev_type = self.dev_type
 
-        Args:
-            payload(bytes): Data to send.
-        """
-        try:
-            s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
-            s.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
-            s.settimeout(self.connection_timeout)
-            s.connect((self.address, self.port))
-        except Exception as e:
-            print("Failed to connect to %s. Raising Exception." % (self.address))
-            raise e
-        try:
+        with socketcontext(self.address, self.port, self.connection_timeout) as s:
             s.send(payload)
-        except Exception as e:
-            print("Failed to send payload to %s. Raising Exception." % (self.address))
-            # s.close()
-            raise e
-
-        try:
             data = s.recv(1024)
-            #            print("FIRST:  Received %d bytes" % len(data) )
+
             # sometimes the first packet does not contain data (typically 28 bytes):
             # need to read again
             if len(data) < 40:
                 time.sleep(0.1)
                 data = s.recv(1024)
-        #                print("SECOND: Received %d bytes" % len(data) )
-        except Exception as e:
-            print("Failed to receive data from %s. Raising Exception." % (self.address))
-            # s.close()
-            raise e
 
-        s.close()
-        return data
+            msg = unpack_message(data)
+            # TODO: Verify stuff, e.g. CRC sequence number
+
+            payload = self._decode_payload(msg.payload)
+
+        # Perform a new exchange (once) if we switched device type
+        if dev_type != self.dev_type:
+            _LOGGER.debug(
+                "Re-send %s due to device type change (%s -> %s)",
+                command,
+                dev_type,
+                self.dev_type,
+            )
+            return self.exchange(command, data)
```

Comment:
This is giving problems, since you are using the same variable (data) that is used to receive the data from the device, so if I print out the json_data of the payload I get:
`DEBUG:localtuya.pytuya:json_data={'devId': 'bf4ce0d8be2bc62e2fgdoy', 'uid': 'bf4ce0d8be2bc62e2fgdoy', 't': '1601443774', 'dps': b'\x00\x00U\xaa\x00\x00\x00\x00\x00\x00\x00\n\x00\x00\x00,\x00\x00\x00\x01\x82\x86>el\xe2\xd1\xe5\xe90(\x05\x9b\xc8\xea\x8aA\xd1\xfb\xc0\x98)Z\x1e\x8er\x05dd?5UTH^\xc4\x00\x00\xaaU'}`

Reply:
Ah, yes, that's not very good.